service/stores.ml

```diff
@@ -0,0 +1,182 @@
+open Eio.Std
+
+module Store = Git_unix.Store
+module Store_map = Map.Make(String)
+
+let git_lock = Eio.Mutex.create ()
+
+type commit = string * string
+
+type t = {
+  cache_dir : string;
+  stores_lock : Eio.Mutex.t;
+  mutable stores : Store.t Store_map.t;
+  process_mgr : Eio.Process.mgr;
+  mutable index_cache : (commit list * Git_context.index) option;
+}
+
+let git_command ?cwd args =
+  "git" ::
+  match cwd with
+  | Some dir -> "-C" :: dir :: args
+  | None -> args
+
+let run_git ?cwd t args =
+  Eio.Process.run t.process_mgr (git_command ?cwd args)
+
+let line_opt r =
+  if Eio.Buf_read.at_end_of_input r then None
+  else Some (Eio.Buf_read.line r)
+
+let run_git_line ?cwd t args =
+  Eio.Process.parse_out t.process_mgr line_opt (git_command ?cwd args)
+
+module Git_clone = struct
+  let replace_special =
+    String.map @@ function
+    | 'A'..'Z'
+    | 'a'..'z'
+    | '0'..'9'
+    | '-' as c -> c
+    | _ -> '_'
+
+  let rec mkdir_p path =
```

Comment:
This is such a useful function we should probably add it to `eio` (like Python's `os.makedirs(..., exist_ok=True)`